pkg/server/hostservices/identityprovider/identityprovider.go

```diff
@@ -0,0 +1,107 @@
+package identityprovider
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"sync"
+
+	"github.com/spiffe/spire/proto/spire/server/datastore"
+	"github.com/spiffe/spire/proto/spire/server/hostservices"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type X509Identity struct {
+	CertChain  []*x509.Certificate
+	PrivateKey crypto.PrivateKey
+}
+
+type X509IdentityFetcher interface {
+	FetchX509Identity(context.Context) (*X509Identity, error)
+}
+
+type X509IdentityFetcherFunc func(context.Context) (*X509Identity, error)
+
+func (fn X509IdentityFetcherFunc) FetchX509Identity(ctx context.Context) (*X509Identity, error) {
+	return fn(ctx)
+}
+
+type Config struct {
+	// TrustDomainID of the server trust domain.
+	TrustDomainID string
+}
+
+type Deps struct {
+	// DataStore is used to retrieve the lastest bundle. It MUST be set.
+	DataStore datastore.DataStore
+
+	// X509IdentityFetcher is used to fetch the X509 identity. It MUST be set.
+	X509IdentityFetcher X509IdentityFetcher
+}
+
+type IdentityProvider struct {
+	config Config
+
+	mu   sync.RWMutex
+	deps *Deps
+}
+
+func New(config Config) *IdentityProvider {
+	return &IdentityProvider{
+		config: config,
+	}
+}
+
+func (s *IdentityProvider) SetDeps(deps Deps) {
+	s.mu.Lock()
+	s.deps = &deps
+	s.mu.Unlock()
+}
+
+func (s *IdentityProvider) getDeps() (*Deps, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	if s.deps == nil {
+		return nil, status.Error(codes.FailedPrecondition, "IdentityProvider host service has not been initialized")
+	}
+	return s.deps, nil
```

Comment:
since DataStore and X509IdentityFetcher must be set in order to really be able to use s.deps, we may check if they are not nil also.

Reply:
👍